source/common/upstream/subset_lb.cc

```diff
@@ -330,8 +333,25 @@ bool SubsetLoadBalancer::hostMatches(const SubsetMetadata& kvs, const Host& host
       return false;
     }
 
-    if (!ValueUtil::equal(entry_it->second, kv.second)) {
-      return false;
+    // If configured to do so, we attempt to match the metadata value with the contents
+    // of the list specified on the host.
+    if (list_as_any_ && entry_it->second.kind_case() == ProtobufWkt::Value::kListValue &&
+        kv.second.kind_case() == ProtobufWkt::Value::kListValue) {
```

Comment:
I'm not sure this is right. 

`findOrCreateSubsets` explodes list metadata into a subset for each list value and `hostMatches` is used as a predicate to check whether a given host belongs in the subset. But I think this implementation means that any of the subsets created for the N list values will match any host that has any of the list values?

If that's correct (and it's possible I'm missing something), then maybe the way out is to move the list exploding to extractSubsetMetadata? Have it return a sequence of kvs and then treat each one separately (e.g. findOrCreateSubsets reverts to its original logic and hostMatches checks host's list values against a kvs that is known not to have any list values).

Reply:
Ah yeah I think you're right. Keeping the list value in the kvs means that the kvs for each subset generated by the explosion have the same kvs. 

I like your suggestion, I'll give that a go (+ add a test that catches this issue).